### ISSUE TITLE
Tensorflow 2.3.0 Java API on Windows - UnsatisfiedLinkError: Cannot find TensorFlow native library for OS

### ISSUE THREAD

**Issue by FreddyLab** [tensorflow/tensorflow#42891]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 8.1 Pro (64bit)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.3.0
- Python version: N/A (using Tensorflow Java API for Windows)
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A (CPU-only)

**Describe the current behavior**
Currently I've problems to run the `HelloTensorFlow.java` with Tensorflow Java 2.3.0 API on Windows. 

The start of `HelloTensorFlow.java` fails with the following error message:
```
org.tensorflow.NativeLibrary: jniResourceName: org/tensorflow/native/windows-x86_64/tensorflow_jni.dll
org.tensorflow.NativeLibrary: frameworkResourceName: org/tensorflow/native/windows-x86_64/tensorflow_framework.dll
Exception in thread "main" java.lang.UnsatisfiedLinkError: Cannot find TensorFlow native library for OS: windows, architecture: x86_64. See https://github.com/tensorflow/tensorflow/tree/master/tensorflow/java/README.md for possible solutions (such as building the library from source). Additional information on attempts to find the native library can be obtained by adding org.tensorflow.NativeLibrary.DEBUG=1 to the system properties of the JVM.
	at org.tensorflow.NativeLibrary.load(NativeLibrary.java:79)
	at org.tensorflow.TensorFlow.init(TensorFlow.java:67)
	at org.tensorflow.TensorFlow.<clinit>(TensorFlow.java:82)
	at org.tensorflow.Graph.<clinit>(Graph.java:479)
	at com.test.HelloTensorFlow.main(HelloTensorFlow.java:12)
```

I double-checked that _-Djava.library.path_ contains the correct path to _tensorflow_jni.dll_. _Visual C++2015 Redistributable_ is also installed on my Windows system.

I've tested the same `HelloTensorFlow.java` on macOS (with the corresponding macOS TF 2.3.0 native libs). Here everything works like a charm.

**Standalone code to reproduce the issue**
I've used 
- `tensorflow_jni.dll` downloaded at https://storage.googleapis.com/tensorflow/libtensorflow/libtensorflow_jni-cpu-windows-x86_64-2.3.0.zip
- `libtensorflow-2.3.0.jar` downloaded at https://storage.googleapis.com/tensorflow/libtensorflow/libtensorflow-2.3.0.jar
- `HelloTensorFlow.java` (copied from https://www.tensorflow.org/install/lang_java)

I've also added the necessary VM argument: -Djava.library.path=<path-to-tensorflow_jni.dll>

**Comment by Saduf2019** [comment]
@FreddyLab 
Please refer to below linksand let us know if it helps: 
[link](https://stackoverflow.com/questions/46303270/exception-in-thread-main-java-lang-unsatisfiedlinkerror-cannot-find-tensorflo), [link1](https://xbuba.com/questions/46303270)

**Comment by FreddyLab** [comment]
@Saduf2019 
Thanks for the quick reply and the links. I've checked them but unfortunately these hints do not help to solve my problem. 
I think my configuration is correct because if I replace the Windows TF Java 2.3.0 libs with older TF Java 1.14.0 libs (tensorflow_jni.dll + libtensorflow-1.14.0.jar) I can run the `HelloTensorFlow.java` without any problem:
```
org.tensorflow.NativeLibrary: isLoaded: true
org.tensorflow.NativeLibrary: isLoaded: true
2020-09-03 09:39:59.861845: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
Hello from 1.14.0
```

The `UnsatisfiedLinkError` ONLY occurs when using current Java TF 2.3.0 API libs (Windows).

**Comment by gowthamkpr** [comment]
@FreddyLab Is the error occuring with tensorflow 2.2.0 or tf-nightly?

**Comment by FreddyLab** [comment]
@gowthamkpr I've only tested with Windows TF Java 2.3.0 libs so far (downloaded at https://www.tensorflow.org/install/lang_java#tensorflow_with_the_jdk). 

I haven't tested with TF 2.2 or tf-nightly yet, because I haven't found any links to download these Libtensorflow JNI packages versions. At the URL https://storage.googleapis.com/libtensorflow-nightly I only found libs for Linux and MacOs but none for Windows. Do you know where I can download TF 2.2 or nightly libtensorflow versions for Windows?

**Comment by gunan** [comment]
Please redirect to TF java owners. I have no experience with our java releases.
For libtensorflow, please reach out to @av8ramit 
For java, I do not know who would be responsible at the moment.

**Comment by av8ramit** [comment]
> I haven't tested with TF 2.2 or tf-nightly yet, because I haven't found any links to download these Libtensorflow JNI packages versions. At the URL https://storage.googleapis.com/libtensorflow-nightly I only found libs for Linux and MacOs but none for Windows. Do you know where I can download TF 2.2 or nightly libtensorflow versions for Windows?

That's odd I don't know why you cannot see it in the browser window, but they are there. Might be a listing issue. `gsutil` might list them better.
Either way, here are the ones from 09/17.
[CPU](https://storage.googleapis.com/libtensorflow-nightly/prod/tensorflow/release/windows/cpu_libtensorflow/nightly/891/20200917-043607/github/tensorflow/lib_package/libtensorflow-cpu-windows-x86_64.zip)
[CPU JNI](https://storage.googleapis.com/libtensorflow-nightly/prod/tensorflow/release/windows/cpu_libtensorflow/nightly/891/20200917-043607/github/tensorflow/lib_package/libtensorflow_jni-cpu-windows-x86_64.zip)
[GPU](https://storage.googleapis.com/libtensorflow-nightly/prod/tensorflow/release/windows/gpu_libtensorflow/nightly/944/20200917-042911/github/tensorflow/lib_package/libtensorflow-gpu-windows-x86_64.zip)
[GPU JNI](https://storage.googleapis.com/libtensorflow-nightly/prod/tensorflow/release/windows/gpu_libtensorflow/nightly/944/20200917-042911/github/tensorflow/lib_package/libtensorflow_jni-gpu-windows-x86_64.zip)

**Comment by FreddyLab** [comment]
> Either way, here are the ones from 09/17.
> [CPU](https://storage.googleapis.com/libtensorflow-nightly/prod/tensorflow/release/windows/cpu_libtensorflow/nightly/891/20200917-043607/github/tensorflow/lib_package/libtensorflow-cpu-windows-x86_64.zip)
> [CPU JNI](https://storage.googleapis.com/libtensorflow-nightly/prod/tensorflow/release/windows/cpu_libtensorflow/nightly/891/20200917-043607/github/tensorflow/lib_package/libtensorflow_jni-cpu-windows-x86_64.zip)
> [GPU](https://storage.googleapis.com/libtensorflow-nightly/prod/tensorflow/release/windows/gpu_libtensorflow/nightly/944/20200917-042911/github/tensorflow/lib_package/libtensorflow-gpu-windows-x86_64.zip)
> [GPU JNI](https://storage.googleapis.com/libtensorflow-nightly/prod/tensorflow/release/windows/gpu_libtensorflow/nightly/944/20200917-042911/github/tensorflow/lib_package/libtensorflow_jni-gpu-windows-x86_64.zip)

Thanks for the links. I have tested both the nightly CPU JNI and GPU JNI version. Unfortunately I still have the same "Can't find dependent libraries" problem. After starting `HelloTensorflow.java` with the VM parameters `-Djava.library.path=jni -Dorg.tensorflow.NativeLibrary.DEBUG=1` I still get the error message:

```
org.tensorflow.NativeLibrary: tryLoadLibraryFailed: C:\seu\workspace\Tensorflow2JavaAPITest\jni\tensorflow_jni.dll: Can't find dependent libraries
org.tensorflow.NativeLibrary: jniResourceName: org/tensorflow/native/windows-x86_64/tensorflow_jni.dll
org.tensorflow.NativeLibrary: frameworkResourceName: org/tensorflow/native/windows-x86_64/tensorflow_framework.dll
Exception in thread "main" java.lang.UnsatisfiedLinkError: Cannot find TensorFlow native library for OS: windows, architecture: x86_64. See https://github.com/tensorflow/tensorflow/tree/master/tensorflow/java/README.md for possible solutions (such as building the library from source). Additional information on attempts to find the native library can be obtained by adding org.tensorflow.NativeLibrary.DEBUG=1 to the system properties of the JVM.
	at org.tensorflow.NativeLibrary.load(NativeLibrary.java:79)
	at org.tensorflow.TensorFlow.init(TensorFlow.java:67)
	at org.tensorflow.TensorFlow.<clinit>(TensorFlow.java:82)
	at org.tensorflow.Graph.<clinit>(Graph.java:479)
	at com.HelloTensorFlow.main(HelloTensorFlow.java:12)
```

If I interpret this error message correctly tensorflow_jni.dll is found but apparently there is another dependency which is missing. Do you have any ideas?

Thx for your help!

**Comment by FreddyLab** [comment]
@av8ramit 
Ignore my last comment. Finally I solved my problem. After I installed Microsoft Visual C++ Redistributable **2019** the problem disappeared. Before that I had only installed the 2015 version.

This means that the TF 2.x JNI libs just work perfectly.However, the documentation on https://www.tensorflow.org/install/lang_java?hl=en#download is not correct/up to date:
> Note: On Windows, the native library (tensorflow_jni.dll) requires msvcp140.dll at runtime. See the Windows build from source guide to install the Visual C++ 2015 Redistributable.
